### PR TITLE
[include-cleaner] Report refs for enum constants used through namespace aliases

### PR DESCRIPTION
**Pull request by kadircet**

### CODE REVIEW

File: clang-tools-extra/include-cleaner/unittests/WalkASTTest.cpp [hunk 1]
```diff
@@ -534,6 +534,8 @@ TEST(WalkAST, Enums) {
   testWalk(R"(namespace ns { enum E { A = 42 }; }
               struct S { using ns::E::A; };)",
            "int e = S::^A;");
+  testWalk(R"(namespace ns { enum E { $explicit^A = 42 }; })",
+           "namespace z = ns; int e = z::^A;");
```
Comment (ilya-biryukov) [inline clang-tools-extra/include-cleaner/unittests/WalkASTTest.cpp:538]: Could you add a test for the global specifier too?
```::A``` has a slightly different representation as `NestedNameSpecifier`, so we might want to make sure this case is also tested. (I **think** the existing code should work just fine, though).

**Review by ilya-biryukov — APPROVED** [review]
LGTM. See a small suggestion about adding another test too.

File: clang-tools-extra/include-cleaner/unittests/WalkASTTest.cpp (hunk 1 above)
Comment (kadircet) [reply clang-tools-extra/include-cleaner/unittests/WalkASTTest.cpp:538]: ugh, apparently for global namespaces, we don't store any decls. hence we were failing. tweaked the code a little.

File: clang-tools-extra/include-cleaner/lib/WalkAST.cpp [hunk 2]
```diff
@@ -146,9 +148,24 @@ class ASTWalker : public RecursiveASTVisitor<ASTWalker> {
...
+      case NestedNameSpecifier::Namespace:
+      case NestedNameSpecifier::NamespaceAlias:
+      case NestedNameSpecifier::Global:
+        return true;
+      case NestedNameSpecifier::TypeSpec:
+      case NestedNameSpecifier::TypeSpecWithTemplate:
+      case NestedNameSpecifier::Super:
+      case NestedNameSpecifier::Identifier:
+        return false;
+      }
```
Comment (ilya-biryukov) [inline clang-tools-extra/include-cleaner/lib/WalkAST.cpp:165]: NIT: maybe add `llvm_unreachable` to make sure we get some error messages if someone adds a new kind of nested-specifier in the future?

File: clang-tools-extra/include-cleaner/lib/WalkAST.cpp [hunk 3]
```diff
@@ -146,9 +148,24 @@ class ASTWalker : public RecursiveASTVisitor<ASTWalker> {
     //
     // If it's an enum constant, it must be due to prior decl. Report references
     // to it when qualifier isn't a type.
-    if (llvm::isa<EnumConstantDecl>(FD)) {
-      if (!DRE->getQualifier() || DRE->getQualifier()->getAsNamespace())
-        report(DRE->getLocation(), FD);
+    auto QualifierIsNamepsaceOrNone = [&DRE]() {
```
Comment (ilya-biryukov) [inline clang-tools-extra/include-cleaner/lib/WalkAST.cpp:151]: NIT: Maybe make this a helper function instead?
Having a lambda reduces the scope of visibility, but we also loose the benefits of making the body of our function smaller. I tend to prefer smaller functions, but up to you.

File: clang-tools-extra/include-cleaner/unittests/WalkASTTest.cpp (hunk 1 above)
Comment (ilya-biryukov) [reply clang-tools-extra/include-cleaner/unittests/WalkASTTest.cpp:538]: so it was a good test case! thanks, LG!

**Review by ilya-biryukov — APPROVED** [review]
Still LGTM. See one minor comment about the code style, but up to you if you want to apply it or not.

File: clang-tools-extra/include-cleaner/lib/WalkAST.cpp (hunk 2 above)
Comment (hokein) [reply clang-tools-extra/include-cleaner/lib/WalkAST.cpp:165]: or even just `default: return false;`.

**Review by hokein — APPROVED** [review]

File: clang-tools-extra/include-cleaner/lib/WalkAST.cpp (hunk 2 above)
Comment (kadircet) [reply clang-tools-extra/include-cleaner/lib/WalkAST.cpp:165]: we already have `Wswitch` in LLVM so there should be warnings (unless we actually add a default clause). added llvm_unreachable to have some more useful error messages at runtime though.

Comment (ilya-biryukov) [reply clang-tools-extra/include-cleaner/lib/WalkAST.cpp]: Yeah, we don't use `-Werror`, though, and don't use the same compiler.
So missing the error during development is not unheard of, and I don't think our precommit CI catches that either.

This is why `llvm_unreachable` still makes sense from my perspective, I always put it as an extra layer of defense.

As for `default: ... `, we have an explicit [rule in LLVM Style Guide](https://llvm.org/docs/CodingStandards.html#don-t-use-default-labels-in-fully-covered-switches-over-enumerations) that asks not to do that for fully-covered switches.

**Pull request merged**